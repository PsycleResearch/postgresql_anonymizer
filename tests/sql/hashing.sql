-- This test cannot be run in a single transcation
-- This test must be run on a database named 'contrib_regression'

CREATE EXTENSION IF NOT EXISTS anon CASCADE;

-- Init
SELECT anon.load();

-- Using the automatic salt
SELECT anon.hash(NULL) IS NULL;
SELECT anon.hash('abcd') = anon.hash('abcd','sha512');

-- With a random salt
SELECT anon.random_hash('abcd') != anon.random_hash('abcd');

-- Restore a predifened salt
SELECT anon.set_secret_salt('4a6821d6z4e33108gg316093e6182b803d0361');

-- Return value is always
SELECT anon.hash('abcd') = '26909982853aafcd56410ae33b70541ee8ed868d7438557f41b28ef062ec22c080ff89a786adde1e2797875e9fa20219ee4f7f076a1f9fc5264241e28a31035b';
SELECT anon.hash('abcd','md5') = 'bb01b7484406325bfa8df2e2f2f8d8fd';
SELECT anon.hash('abcd','sha224') = 'f75dd47c245113ab5734270ac13db1f73f330179b10e57aa6b46457b';
SELECT anon.hash('abcd','sha256') = '3a878e51ae30c795fc04554096c6c05d50771f3c6f7352d950bbba6538892346';
SELECT anon.hash('abcd','sha384') = 'd145dd187e478b2fabbe687c119954d9f166a9461319e300ac51ad2ec5ee54a71f39e0ef654ff20098efadf40e5df394';

-- Dynamic masking
SELECT anon.start_dynamic_masking();

CREATE TABLE phone (
  phone_owner  TEXT,
  phone_number TEXT
);

INSERT INTO phone VALUES
('Omar Little','410-719-9009'),
('Russell Bell','410-617-7308'),
('Avon Barksdale','410-385-2983');

CREATE TABLE phonecall (
  call_id INT,
  call_sender TEXT,
  call_receiver TEXT,
  call_start_time TIMESTAMP WITH TIME ZONE,
  call_end_time TIMESTAMP WITH TIME ZONE
);

INSERT INTO phonecall VALUES
(834,'410-617-7308','410-385-2983','2004-05-17 09:41:01.859137+00','2004-05-17 09:44:24.119237+00'),
(835,'410-385-2983','410-719-9009','2004-05-17 11:22:51.859137+00','2004-05-17 11:34:18.119237+00');

SECURITY LABEL FOR anon ON COLUMN phone.phone_owner
IS 'MASKED WITH FUNCTION concat(anon.pseudo_first_name(phone_owner),$$ $$,anon.pseudo_last_name(phone_owner))';

SECURITY LABEL FOR anon ON COLUMN phone.phone_number
IS 'MASKED WITH FUNCTION anon.hash(phone_number)';

SECURITY LABEL FOR anon ON COLUMN phonecall.call_sender
IS 'MASKED WITH FUNCTION anon.hash(call_sender)';

SECURITY LABEL FOR anon ON COLUMN phonecall.call_receiver
IS 'MASKED WITH FUNCTION anon.hash(call_receiver)';


-- ROLE

CREATE ROLE jimmy_mcnulty LOGIN;

SECURITY LABEL FOR anon ON ROLE jimmy_mcnulty IS 'MASKED';

SELECT anon.mask_update();

-- Jimmy reads the phone book
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT * FROM phone'

-- Jimmy joins table to get the call history
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT p1.phone_owner as "from", p2.phone_owner as "to", c.call_start_time FROM phonecall c JOIN phone p1 ON c.call_sender = p1.phone_number JOIN phone p2 ON c.call_receiver = p2.phone_number'

-- Jimmy tries to find the salt :-)
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT * FROM anon.secret'
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT anon.get_secret_salt();'



-- STOP

SELECT anon.stop_dynamic_masking();

--  CLEAN

DROP TABLE phonecall CASCADE;
DROP TABLE phone CASCADE;

DROP EXTENSION anon CASCADE;

REASSIGN OWNED BY jimmy_mcnulty TO postgres;
DROP OWNED BY jimmy_mcnulty CASCADE;
DROP ROLE jimmy_mcnulty;

