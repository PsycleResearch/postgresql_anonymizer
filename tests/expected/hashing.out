-- This test cannot be run in a single transcation
-- This test must be run on a database named 'contrib_regression'
CREATE EXTENSION IF NOT EXISTS anon CASCADE;
-- Init
SELECT anon.load();
 load 
------
 t
(1 row)

-- Using the automatic salt
SELECT anon.hash(NULL) IS NULL;
 ?column? 
----------
 t
(1 row)

SELECT anon.hash('abcd') = anon.hash('abcd','sha512');
 ?column? 
----------
 t
(1 row)

-- With a random salt
SELECT anon.random_hash('abcd') != anon.random_hash('abcd');
 ?column? 
----------
 t
(1 row)

-- Restore a predifened salt
SELECT anon.set_secret_salt('4a6821d6z4e33108gg316093e6182b803d0361');
            set_secret_salt             
----------------------------------------
 4a6821d6z4e33108gg316093e6182b803d0361
(1 row)

-- Return value is always
SELECT anon.hash('abcd') = '26909982853aafcd56410ae33b70541ee8ed868d7438557f41b28ef062ec22c080ff89a786adde1e2797875e9fa20219ee4f7f076a1f9fc5264241e28a31035b';
 ?column? 
----------
 t
(1 row)

SELECT anon.hash('abcd','md5') = 'bb01b7484406325bfa8df2e2f2f8d8fd';
 ?column? 
----------
 t
(1 row)

SELECT anon.hash('abcd','sha224') = 'f75dd47c245113ab5734270ac13db1f73f330179b10e57aa6b46457b';
 ?column? 
----------
 t
(1 row)

SELECT anon.hash('abcd','sha256') = '3a878e51ae30c795fc04554096c6c05d50771f3c6f7352d950bbba6538892346';
 ?column? 
----------
 t
(1 row)

SELECT anon.hash('abcd','sha384') = 'd145dd187e478b2fabbe687c119954d9f166a9461319e300ac51ad2ec5ee54a71f39e0ef654ff20098efadf40e5df394';
 ?column? 
----------
 t
(1 row)

-- Dynamic masking
SELECT anon.start_dynamic_masking();
 start_dynamic_masking 
-----------------------
 t
(1 row)

CREATE TABLE phone (
  phone_owner  TEXT,
  phone_number TEXT
);
INSERT INTO phone VALUES
('Omar Little','410-719-9009'),
('Russell Bell','410-617-7308'),
('Avon Barksdale','410-385-2983');
CREATE TABLE phonecall (
  call_id INT,
  call_sender TEXT,
  call_receiver TEXT,
  call_start_time TIMESTAMP WITH TIME ZONE,
  call_end_time TIMESTAMP WITH TIME ZONE
);
INSERT INTO phonecall VALUES
(834,'410-617-7308','410-385-2983','2004-05-17 09:41:01.859137+00','2004-05-17 09:44:24.119237+00'),
(835,'410-385-2983','410-719-9009','2004-05-17 11:22:51.859137+00','2004-05-17 11:34:18.119237+00');
SECURITY LABEL FOR anon ON COLUMN phone.phone_owner
IS 'MASKED WITH FUNCTION concat(anon.pseudo_first_name(phone_owner),$$ $$,anon.pseudo_last_name(phone_owner))';
SECURITY LABEL FOR anon ON COLUMN phone.phone_number
IS 'MASKED WITH FUNCTION anon.hash(phone_number)';
SECURITY LABEL FOR anon ON COLUMN phonecall.call_sender
IS 'MASKED WITH FUNCTION anon.hash(call_sender)';
SECURITY LABEL FOR anon ON COLUMN phonecall.call_receiver
IS 'MASKED WITH FUNCTION anon.hash(call_receiver)';
-- ROLE
CREATE ROLE jimmy_mcnulty LOGIN;
SECURITY LABEL FOR anon ON ROLE jimmy_mcnulty IS 'MASKED';
SELECT anon.mask_update();
 mask_update 
-------------
 t
(1 row)

-- Jimmy reads the phone book
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT * FROM phone'
  phone_owner  |                                                           phone_number                                                           
---------------+----------------------------------------------------------------------------------------------------------------------------------
 uzziah Vickey | 3fddbf5071a02d45aba1c5760a05f8183b8fedfa49e79c65f87cee3c4de6646f3c4a477ae0f5fbeec168e1f548f92185ffe73f6863e3294a742d75515cca1dab
 ábel Aburto   | ebf4db3d96dc4df7c5ef63a1d005136dd3700a8ba161c5ad874f8922323bf0ba56946c93c10916e30900636f51db073f35f2b93ccc52edbd5e80e5231ab97540
 martín Movlin | 420287d783a7b5271da53d19b0a96c0c1eeaf17b9052bfabc74a61696a5be4e22333b6c1596a3d1e4ab68d6a1784de8973d85af37a2a7eaf3f225cb795da8806
(3 rows)

-- Jimmy joins table to get the call history
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT p1.phone_owner as "from", p2.phone_owner as "to", c.call_start_time FROM phonecall c JOIN phone p1 ON c.call_sender = p1.phone_number JOIN phone p2 ON c.call_receiver = p2.phone_number'
     from      |      to       |           call_start_time           
---------------+---------------+-------------------------------------
 martín Movlin | uzziah Vickey | Mon May 17 04:22:51.859137 2004 PDT
 ábel Aburto   | martín Movlin | Mon May 17 02:41:01.859137 2004 PDT
(2 rows)

-- Jimmy tries to find the salt :-)
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT * FROM anon.secret'
ERROR:  permission denied for table secret
\! psql contrib_regression -U jimmy_mcnulty -c 'SELECT anon.get_secret_salt();'
ERROR:  permission denied for function get_secret_salt
-- STOP
SELECT anon.stop_dynamic_masking();
NOTICE:  The previous priviledges of 'jimmy_mcnulty' are not restored. You need to grant them manually.
 stop_dynamic_masking 
----------------------
 t
(1 row)

--  CLEAN
DROP TABLE phonecall CASCADE;
DROP TABLE phone CASCADE;
DROP EXTENSION anon CASCADE;
REASSIGN OWNED BY jimmy_mcnulty TO postgres;
DROP OWNED BY jimmy_mcnulty CASCADE;
DROP ROLE jimmy_mcnulty;
